src/libraries/System.Private.CoreLib/src/System/Numerics/Decimal128.cs

```diff
@@ -0,0 +1,362 @@
+            {
+                return (value is null) ? 1 : throw new ArgumentException(SR.Arg_MustBeDecimal128);
+            }
+            return CompareTo(other);
+        }
+
+        /// <inheritdoc cref="IComparable{T}.CompareTo(T)" />
+        public int CompareTo(Decimal128 other)
+        {
+            var current = new UInt128(_upper, _lower);
+            var another = new UInt128(other._upper, other._lower);
+            return Number.CompareDecimalIeee754<Decimal128, UInt128>(current, another);
+        }
+
+        /// <inheritdoc cref="IEquatable{T}.Equals(T)" />
+        public bool Equals(Decimal128 other)
+        {
+            var current = new UInt128(_upper, _lower);
+            var another = new UInt128(other._upper, other._lower);
+            return Number.CompareDecimalIeee754<Decimal128, UInt128>(current, another) == 0;
+        }
+
+        /// <summary>
+        /// Returns a value that indicates whether this instance is equal to a specified <paramref name="obj"/>.
+        /// </summary>
+        public override bool Equals([NotNullWhen(true)] object? obj)
+        {
+            return obj is Decimal128 && Equals((Decimal128)obj);
+        }
+
+        /// <summary>
+        /// Serves as the default hash function.
+        /// </summary>
+        public override int GetHashCode()
+        {
+            return new UInt128(_upper, _lower).GetHashCode();
+        }
+
+        /// <summary>
+        /// Returns a string representation of the current value.
+        /// </summary>
+        public override string ToString()
+        {
+            return Number.FormatDecimalIeee754<Decimal128, UInt128>(this, null, NumberFormatInfo.CurrentInfo);
+        }
+
+        /// <summary>
+        /// Returns a string representation of the current value using the specified <paramref name="format"/>.
+        /// </summary>
+        public string ToString([StringSyntax(StringSyntaxAttribute.NumericFormat)] string? format)
+        {
+            return Number.FormatDecimalIeee754<Decimal128, UInt128>(this, format, NumberFormatInfo.CurrentInfo);
+        }
+
+        /// <summary>
+        /// Returns a string representation of the current value with the specified <paramref name="provider"/>.
+        /// </summary>
+        public string ToString(IFormatProvider? provider)
+        {
+            return Number.FormatDecimalIeee754<Decimal128, UInt128>(this, null, NumberFormatInfo.GetInstance(provider));
+        }
+
+        /// <summary>
+        /// Returns a string representation of the current value using the specified <paramref name="format"/> and <paramref name="provider"/>.
+        /// </summary>
+        public string ToString([StringSyntax(StringSyntaxAttribute.NumericFormat)] string? format, IFormatProvider? provider)
+        {
+            return Number.FormatDecimalIeee754<Decimal128, UInt128>(this, format, NumberFormatInfo.GetInstance(provider));
+        }
+
+        private static UInt128[] UInt128Powers10 =>
```

Comment:
This can be a problem for constant embedding. `long` array can be encoded into executable and read by pairs.